src/toxcore/dht_new/packet.rs

```diff
@@ -178,6 +179,49 @@ impl FromBytes for DhtPacket {
     ));
 }
 
+impl DhtPacket {
+    /**
+    Decrypt payload and try to parse it as packet type.
+
+    To get info about it's packet type use
+    [`.kind()`](./struct.DhtPacket.html#method.kind) method.
+
+    Returns `Error` in case of failure:
+
+    - fails to decrypt
+    - fails to parse as given packet type
+    */
+    pub fn get_payload(&self, own_secret_key: &SecretKey) -> Result<Option<DhtPacketPayload>, Error>
+    {
+        debug!(target: "DhtPacket", "Getting packet data from DhtPacket.");
+        trace!(target: "DhtPacket", "With DhtPacket: {:?}", self);
+        let decrypted = open(&self.payload, &self.nonce, &self.pk,
+                            own_secret_key)
+            .and_then(|d| Ok(d))
+            .map_err(|e| {
+                debug!("Decrypting DhtPacket failed!");
+                Error::new(ErrorKind::Other,
+                    format!("DhtPacket decrypt error: {:?}", e))
+            });
+
+        match DhtPacketPayload::from_bytes(&decrypted.unwrap_or(vec![0]), self.packet_kind) {
```

Comment:
Why unwrap_or(vec![0])? You already did .map_err(..) part so just add question mark to return error properly.